dakota/src/dom.rs

```diff
@@ -90,56 +90,43 @@ pub struct Content {
     pub el: DakotaId,
 }
 
-/// This is a relative value that modifies an element
-/// by a percentage of the size of the available space.
+/// Represents a possibly relative value. This will
+/// either be a f32 scaling value or a constant size
+/// u32.
 #[derive(Debug, PartialEq, PartialOrd, Copy, Clone)]
-pub struct Relative {
-    val: f32,
+pub enum Value {
+    /// This is a relative value that modifies an element
+    /// by a percentage of the size of the available space.
+    Relative(f32),
+    Constant(i32),
 }
 
-impl Relative {
-    pub fn new(v: f32) -> Self {
-        assert!(v >= 0.0 && v <= 1.0);
-        Self { val: v }
-    }
-
-    pub fn scale(&self, val: f32) -> Result<u32> {
-        if !(self.val >= 0.0 && self.val <= 1.0) {
+impl Value {
+    fn scale(current: f32, val: f32) -> Result<u32> {
+        if !(current >= 0.0 && current <= 1.0) {
             return Err(anyhow!(
                 "Element.relativeOffset should use values in the range (0.0, 1.0)"
             ));
         }
-        Ok((val * self.val) as u32)
+        Ok((current * val) as u32)
     }
-}
 
-#[derive(Debug, PartialEq, PartialOrd, Copy, Clone)]
-pub struct Constant {
-    val: i32,
-}
-
-impl Constant {
-    pub fn new(val: i32) -> Self {
-        Self { val: val }
-    }
-}
-
-/// Represents a possibly relative value. This will
-/// either be a f32 scaling value or a constant size
-/// u32.
-#[derive(Debug, PartialEq, PartialOrd, Copy, Clone)]
-pub enum Value {
-    Relative(Relative),
-    Constant(Constant),
-}
-
-impl Value {
     pub fn get_value(&self, avail_space: f32) -> Result<i32> {
-        Ok(match self {
-            Self::Relative(r) => r.scale(avail_space)? as i32,
-            Self::Constant(c) => c.val,
+        Ok(match *self {
+            Self::Relative(r) => Self::scale(r, avail_space)? as i32,
+            Self::Constant(c) => c,
         })
     }
+
+    pub fn new(&self) -> Self {
+        match *self {
+            Self::Relative(r) => {
+                assert!(r >= 0.0 && r <= 1.0);
```

Comment:
I think we should get rid of the `new` method. It looks pretty weird doing `            dom::Value::Constant(resolution.0 as i32).new(),`.

Instead I think we should move this assert into `scale`, and then we can do `dom::Value::Constant(resolution.0 as i32).` and the value check will happen when it's used.

Reply:
Is this check not already happening in scale? Only it's a conditional that throws instead off an assertion?

Reply:
Isn't this already sort of happening in `scale()` via the conditional that throws if true? Would scale be implicitly called on instantiation?

Reply:
Heh I don't even remember  adding that. Then yes that should do it, you can go ahead with removing `new`

Reply:
Removing new breaks the `viewports.xml` file. Investigating this now.